### ISSUE TITLE
Onboarding modal looping?

### ISSUE THREAD

**Issue by yheuhtozr** [forem/forem#14483]
**Describe the bug**

I have an instance for testing purpose, and a few users (authenticated via Twitter) reported that they experienced the onboarding modal looping forever even if they have followed all the steps. Now when I signed up https://forem.dev with my GitHub account, it actually happened to me!

When my users reported the bug, I confirmed that their `saw_onboarding`, `checked_code_of_conduct`, and `checked_terms_and_conditions` columns in the user table were false. I guess that so is my record (username: `yheuhtozr`) on forem.dev now. I have no idea whether I can prove that I have already gone through the onboarding slides (not everyone experiencing this), but would like to report as it is.

**To Reproduce**

1. Sign up an instance using (Twitter|GitHub) (?) for the first time
2. Finish five slides "Welcome to Forem!" - "Email preferences"
3. It goes back to the first slide, all options cleared, after the normal page seems blinked a second

**Expected behavior**

Bring me back to the top page after the onboarding settings.

**Screenshots**

URL box: `https://forem.dev/onboarding?referrer=https://forem.dev/`

![image](https://user-images.githubusercontent.com/84892012/129174468-b64df4a8-a57b-4afb-9986-43cf032bc680.png)

**Desktop (please complete the following information):**

- OS, version: Windows 10 Home 21H1
- Browser, version: Firefox 91.0

**Additional context**

My users also reportedly use Edge or Chrome.

**Comment by cmgorton** [comment]
I'll keep it open for a bit to see if anyone else experiences this issue. If no one else reports it soon we can close this issue 😃

**Comment by citizen428** [comment]
@cmgorton I experience it quite frequently during local development after wiping my DB. This may be worth exploring.

**Comment by yheuhtozr** [comment]
FWIW after some communication with testers of my instance, there is a sign that who experience this have session or auth problem actually. If I tell them to log in from another browser, the problem is gone and never happens again.

**Comment by nickytonline** [comment]
It is highly unlikely that someone would load the onboarding page directly from the browser address bar, but that would be the use case to have both. Otherwise, `InstantClick` on its own is probably fine in this scenario since it's converted the page to a SPA like experience at that point.

Experiment a bit and let me know if you have any other questions. 😎

**Comment by BlueZ-95** [comment]
> It is highly unlikely that someone would load the onboarding page directly from the browser address bar, but that would be the use case to have both. Otherwise, `InstantClick` on its own is probably fine in this scenario since it's converted the page to a SPA like experience at that point.
> 
> Experiment a bit and let me know if you have any other questions. 😎

Owkay, Lemme experiment a bit!

**Comment by LeRayan** [comment]
Do we have any updates on this? I have this bug very often on my forem instance:  www.developmenthackers.com
It happens on both Brave browser and Chrome. I've noticed that on the first onboarding modal, there's a parameter in the URL which says referrer=none, and when it loops again it says referrer=https://www.developmenthackers.com/

I would appreciate any updates on this as I'm kind of waiting for this fix before launching the community. I've also tried updating and relaunching my forem instance.
Thank you :)

**Comment by djuber** [comment]
exploratory notes

I definitely observed a few times (today and yesterday) when testing on dev.to, and once yesterday on benhalperncommunity

When repeating the onboarding, currentUser has `saw_onboarding: false` (as does the localStorage and document.body.dataset.user copies). 

I feel like we ruled out the possibility of a failed user save (the users controller would respond with a 422 if we didn't persist the checkbox and onboarding flag) - the onboarding dialog would not show the next slide (tag follows) if the response was not 200 - we changed the onboarding checkbox update method to pretend the save had failed and confirmed a 200 was not returned, a 422 was seen in the network logs, and the submit task did not move to the next slide in the onboarding process.
 
What we believe we saw was that the request to /async_info/base_data after the onboarding returned a user with saw_onboarding set to false.

suspected sources in order of likelihood
- dirty write to the users table happening in a background process from a callback, or a separate request, unsetting saw_onboarding and the checkbox fields (are there background jobs enqueued when a user is created, that save the object and overwrite the ?)
- stale currentUser being used from the localStorage object used to decide whether to display the onboarding page (the fix for this would be to persist in the IntroSlide (where we write "false" to the localStorage object 'shouldRedirectToOnboarding') a modified user. One vector for re-using the stale object would be that we load the current_user from localStorage before the fetch response from async_info completes. My understanding of the onboarding redirect check is the call to getUserDataAndCsrfToken should ensure we've waited for currentUser to have been re-assigned from the base data response. I believe we observed the base data response included a false saw_onboarding during testing. I'll repeat testing to focus on that situation more closely to confirm. This possible cause (stale user data being considered) has the benefit of being easily fixed, and the disadvantage of possibly being the wrong problem.

Andy will look further into monitoring (determining how prevalent this is, establishing a base rate, so we can identify whether it has improved after a fix is proposed)).

**Comment by djuber** [comment]
I've isolated this to the csrf protection. Sending the same request with the same cookies with an invalid x-csrf-token header gives the results we're seeing when onboarding is looping (0 byte body, 200 response status).

**Comment by benhalpern** [comment]
The `ValidRequest` hack had its origins in how DEV was misconfigured where SSL was set at the edge, but `  config.force_ssl` was not set to `true`. That's why the code currently reads `ENV["FORCE_SSL_IN_RAILS"] == "true"`... However, this _is_ set to `true` at least on DEV.

However, with Forem Cloud I think it might have actually been appropriate to set this to `false`, or at least I recall being surprised that there was still reason to have this ENV var once we got the problem resolved in DEV. So I'm not sure what the current config is in Forem Cloud related to all of this.

That's as much old context as I can provide on the issue if that helps anyone — whether or not this is _actually the problem_.

>  Sending the same request with the same cookies with an invalid x-csrf-token header gives the results we're seeing when onboarding is looping (0 byte body, 200 response status).

@djuber just for my curiosity, can you expand on what you mean here? Do we know the origin of this "invalid x-csrf-token"? Do we have visibility into whether we're sending _any_ token, or an expired/invalid one?

**Comment by djuber** [comment]
Testing this in a browser, if after seeing the 0 byte response from `PATCH /onboarding_checkbox_update`, I call fetchBaseData(), the response includes a new csrf token, and resending the patch succeeds (this is fixable by refreshing the data from the server).

The global csrfToken variable in the app does not change, but the csrfToken sent in the requests does change

Failing request headers

```
PATCH /onboarding_checkbox_update HTTP/2
Host: dev.to
Accept: */*
Referer: https://dev.to/onboarding?referrer=https%3A%2F%2Fdev.to%2Fusers%2Fsign_up
X-CSRF-Token: 5zmJPDbZEOfFB2Wj2bg...
Content-Type: application/json
Origin: https://dev.to
Content-Length: 77
Connection: keep-alive
```

![console](https://user-images.githubusercontent.com/1237369/163009894-91fe140d-b451-4d32-a273-134113960faf.png)

The base_data respnonse made between those two csrfToken checks returned a new token 

```
"token": "fOgMXQc22Ru..."
```

This was used in the second request which received the expected response from the controller. I'll see what's involved in adding that to the intro slide code (all the subsequent slides in the flow will work fine, since they'll use the same valid token for their updates).

```
PATCH /onboarding_checkbox_update HTTP/2
Host: dev.to
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:100.0) Gecko/20100101 Firefox/100.0
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
Referer: https://dev.to/onboarding?referrer=https%3A%2F%2Fdev.to%2Fusers%2Fsign_up
X-CSRF-Token: fOgMXQc22Ru...
Content-Type: application/json
Origin: https://dev.to
Content-Length: 77
Connection: keep-alive
```

**Comment by amywtlin** [comment]
@Zhao-Andy Are we still actively looking at this one? I'd love to keep this one moving, either to have you look at it (if your bandwidth permits) or you can remove your assignment so other teammates can pick it up.

**Comment by Zhao-Andy** [comment]
@amywtlin yes thanks for the reminder on this! I have a test instance I can use to play around with. Will add this to my to-do list.

**Comment by amywtlin** [comment]
That'll be awesome - thanks so much!

**Comment by michael-tharrington** [comment]
This isn't something that gets reported frequently, but I just wanted to confirm that this is still occurring. I just gave it a test on the Web Monetization Community here:

https://www.loom.com/share/4135bbe514744afcada4f2089e9f6af0?t=206

**Comment by pkfrank** [comment]
In case it helps us troubleshoot, I was able to reproduce this reliably a few times in a row today (I have a Loom if helpful), when trying to register via Google OAuth on...

* `https://www.flowstate.to/enter`
* `https://www.thismmalife.com/enter`

**Comment by fdocr** [comment]
Sharing another update on the issue:

I've seen this happen to me on production instances (fleet hosted instances and not on Heroku ones like [BHC](https://community.benhalpern.com/)) after taking actions through onboarding like following tags and enabling email updates on the last slide. This hasn't been the case when "rushing through the steps" as quickly as possible. That's the closest I've been able to get to reliably recreating the error. I've also been unable to recreate it on local environment despite numerous loops through the onboarding flow.

We have ongoing work on refactoring the onboarding flow, so the decision here is now to lean into the refactor and stop investing time in this bug.

**Comment by mikeydorje** [comment]
Closing due to age and inactivity. The onboarding flow has changed significantly since this was reported. If this is still reproducible, please open a new issue with current steps to reproduce.
